File: Striver_SDE_SHEET/1087-longest-arithmetic-subsequence/longest-arithmetic-subsequence.py
class Solution:
    def longestArithSeqLength(self, nums: List[int]) -> int:
        n = len(nums)
        dp = {}
        for i in range(n):
            dp[i] = {}
            for j in range(i):
                dp[i][nums[i]-nums[j]] = 1+dp[j].get(nums[i]-nums[j],1)
                
        return max(max(dp[i].values()) for i in range(n) if len(dp[i])>0)

File: Striver_SDE_SHEET/1087-longest-arithmetic-subsequence/longest-arithmetic-subsequence.java
class Solution {
    public int longestArithSeqLength(int[] nums) {
        int N = nums.length;
        Map<Integer,Map<Integer,Integer>> dp = new HashMap<>();
        for(int i=0;i<N;i++){
            dp.put(i,new HashMap<Integer,Integer>());
            for(int j=0;j<i;j++){
                dp.get(i).put(nums[i]-nums[j], 1+dp.get(j).getOrDefault(nums[i]-nums[j],1));
            }
        }
        return dp.values().stream().flatMap(m->m.values().stream()).max(Integer::compare).orElse(0);
    }
}